### ISSUE TITLE
Add collaboration support

### ISSUE THREAD

**Issue by nt1m** [PalmerAL/scratchpad#21]
http://togetherjs.com seems like a great candidate.

**Comment by PalmerAL** [comment]
I tried that. Unfortunately, it doesn't work with contenteditable.

**Comment by nt1m** [comment]
@PalmerAL, it does, if you look deeper into the documentation.
Btw, please don't make any changes, I'm about to submit a big PR

**Comment by nt1m** [comment]
@PalmerAL You can use TogetherJS.send() to send data to other computers, and TogetherJS.on("") to listen for data.

**Comment by PalmerAL** [comment]
Okay, that looks good. Also, I think we should decide how documents are saved and stored before we implement stuff like this.

**Comment by nt1m** [comment]
This is also useful : https://hacks.mozilla.org/2015/02/broadcastchannel-api-in-firefox-38/

**Comment by PalmerAL** [comment]
One issue here is that the dropbox api doesn't support sharing documents.

**Comment by nt1m** [comment]
There's no perfect API really. I would personally use togetherjs for collaboration, and use the dropbox api for saving.

**Comment by PalmerAL** [comment]
The problem, though, is how the two interact. You can't share a document because you can't access the same record from two accounts, and syncing between two different copies of the documents gets messy. This might be implemented later, but for now I think it's not a priority.

**Comment by nt1m** [comment]
@PalmerAL TogetherJS is pretty simple to implement. And you won't need to worry about all the interaction stuff, since TogetherJS is a *temporary* collaboration session, so you can invite others to edit a doc (without saving it to the other dropboxes) using a link, then when people are done, it'll only save on the original author's side.

**Comment by PalmerAL** [comment]
That's a great solution!

**Comment by PalmerAL** [comment]
Copied from wiki:

So keeping aside the gorgeous UI of Scratchpad, wouldn't it be awesome to have collaborations? like it could be a place wherein users can create documents and invite people to join editing that document and chat with them in real time. Mozilla has created a new web technology called TogetherJS and it's very easy to install. The project page has some examples, documentation and information that would suffice you if you're hearing about TogetherJS for the first time. What do you all say?

**Comment by nt1m** [comment]
@PalmerAL Is this supposed to be a paragraph in the landing page ? If so, it sounds more like an advertisement for TogetherJS. You should probably insist more on the features TogetherJS would provide.

**Comment by PalmerAL** [comment]
No, it isn't. @haze007 posted this on the wiki as a suggestion, but this issue is a better place for it. I copied it to here so that the wiki page could be removed.

**Comment by PalmerAL** [comment]
The new Dropbox API allows us to generate permanent sharing url's, so for now I think I'm going to implement the ability to generate a shareable link that lets others view (but not edit) the document. This also means we don't have to deal with the script injection issues in the editor right now (since the shared document will be in a sandboxed iframe).

**Comment by PalmerAL** [comment]
Dropbox API sharing links implemented in eabd5ef8972b3a5cd4de686f6f10cad6d9c2ff09.

**Comment by nt1m** [comment]
@PalmerAL Nice ! I would probably add a description to that dialog though.

**Comment by PalmerAL** [comment]
@nt1m what would you suggest?

**Comment by nt1m** [comment]
@PalmerAL Something like : 
Share this document by sending the unique link below

**Comment by PalmerAL** [comment]
This should be completed by 6bafb8b7f859bbda8e80dc4bb56fec5a938a40f5.

**Comment by nt1m** [comment]
@PalmerAL Are you sure this is the right issue ? This about editing collaboration, not UX improvements to the sharing feature.

**Comment by PalmerAL** [comment]
Yes and no. It originally was for collaboration, but it turned into being about the sharing feature. I guess it should be reopened, though, since the original item this issue was about still hasn't been fixed.
